### ISSUE TITLE
Yolov5 classification mean std for TNN and NCNN frameworks

### ISSUE THREAD

**Issue by xellDart** [ultralytics/yolov5#7536]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi @glenn-jocher , thanks for your amazing work, I have one question, I train my dataset using classifier.py, with accuracy of 0.981, later i use this code using ncnn for classification task:

```
std::vector < float > softmax(const float * logits, unsigned int _size, unsigned int & max_id) {

  if (_size == 0 || logits == nullptr) return {};
  float max_prob = 0. f, total_exp = 0. f;
  std::vector < float > softmax_probs(_size);
  for (unsigned int i = 0; i < _size; ++i) {
    softmax_probs[i] = std::exp((float) logits[i]);
    total_exp += softmax_probs[i];
  }
  for (unsigned int i = 0; i < _size; ++i) {
    softmax_probs[i] = softmax_probs[i] / total_exp;
    if (softmax_probs[i] > max_prob) {
      max_id = i;
      max_prob = softmax_probs[i];
    }
  }
  return softmax_probs;
}



std::vector < unsigned int > argsort(const std::vector < float > & arr) {
  if (arr.empty()) return {};
  const unsigned int _size = arr.size();
  std::vector < unsigned int > indices;
  for (unsigned int i = 0; i < _size; ++i) indices.push_back(i);
  std::sort(indices.begin(), indices.end(),
    [ & arr](const unsigned int a,
      const unsigned int b) {
      return arr[a] > arr[b];
    });
  return indices;
}



static float detect_yolo_classifier(const cv::Mat & image) {
    ncnn::Net yolo_classifier;

    yolo_classifier.opt.use_vulkan_compute = false;

    yolo_classifier.load_param("model.param");
    yolo_classifier.load_model("model.bin");

    const int target_size = 352;

    ncnn::Mat in = ncnn::Mat::from_pixels_resize(image.data, ncnn::Mat::PIXEL_BGR2RGB, image.cols, image.rows, target_size, target_size);

    const float mean_vals[3] = {
      0. f,
      0. f,
      0. f
    };
    const float norm_vals[3] = {
      1.0 / 255. f,
      1.0 / 255. f,
      1.0 / 255. f
    };

 in .substract_mean_normalize(mean_vals, norm_vals);

    ncnn::Extractor ex = yolo_classifier.create_extractor();

    ex.input("in0", in );
    ncnn::Mat out;
    ex.extract("out0", out);

    const unsigned int num_classes = out.w;

    const float * logits = (float * ) out.data;
    unsigned int max_id;
    std::vector < float > scores = softmax(logits, num_classes, max_id);
    std::vector < unsigned int > sorted_indices = argsort(scores);

    return sorted_indices[0];
```

Using python original method in yolov5 repo all is ok, but using ncnn with c++  only detect one class correctly, can you check my code  please? I need to use padding image like yolo detector or not?

Like this?

```
// yolov5/models/common.py DetectMultiBackend
    const int max_stride = 64;

    // letterbox pad to multiple of max_stride
    int w = img_w;
    int h = img_h;
    float scale = 1.f;
    if (w > h)
    {
        scale = (float)target_size / w;
        w = target_size;
        h = h * scale;
    }
    else
    {
        scale = (float)target_size / h;
        h = target_size;
        w = w * scale;
    }

    ncnn::Mat in = ncnn::Mat::from_pixels_resize(bgr.data, ncnn::Mat::PIXEL_BGR2RGB, img_w, img_h, w, h);

    // pad to target_size rectangle
    // yolov5/utils/datasets.py letterbox
    int wpad = (w + max_stride - 1) / max_stride * max_stride - w;
    int hpad = (h + max_stride - 1) / max_stride * max_stride - h;
    ncnn::Mat in_pad;
    ncnn::copy_make_border(in, in_pad, hpad / 2, hpad - hpad / 2, wpad / 2, wpad - wpad / 2, ncnn::BORDER_CONSTANT, 114.f);

```
Im  using script to export to torchscript

python3 export.py --weights '/home/miguel/Documentos/yolov5/runs/train/exp/weights/best.pt'  --include torchscript  --img 320

Using torchscripts i get same results with accuracy of 97 - 98 %

I think that the major probles is mean and std in NCNN, 

In classification script use this:

T.Normalize((0.5, 0.5, 0.5), (0.25, 0.25, 0.25))]

What is the correct values to use in NCNN c++?

Thanks

**Comment by leemengwei** [comment]
@xellDart 
Hi ,  as I noticed substracting mean and std cause bad performance.
I assume you shouldn't preprocess your data with mean & std, as yolov5 repo does not perform those preprocess neither (check https://github.com/ultralytics/yolov5/issues/3358#issuecomment-851012420). I want to know if you solve your problem on close.

**Comment by xellDart** [comment]
Not really, my question is more towards an implementation of the classifier in c++ than Yolov5, thanks

**Comment by nihui** [comment]
https://github.com/Tencent/ncnn/blob/3f6fbcc7a2af24265049555715f294886372bf88/examples/yolov5.cpp#L321
https://github.com/Tencent/ncnn/blob/3f6fbcc7a2af24265049555715f294886372bf88/examples/yolov5_pnnx.cpp#L249

yolov5 ncnn

**Comment by xellDart** [comment]
Thanks but im not using yolov5 detector, my problem is with only classifier implementation(https://github.com/ultralytics/yolov5/blob/classifier/classifier.py),

With conversion HWC to CHW and later to BCWH



im = np.ascontiguousarray(np.asarray(im).transpose((2, 0, 1)))  # HWC to CHW
im = torch.tensor(im).float().unsqueeze(0) / 255.0  # to Tensor, to BCWH, rescale
im = resize(normalize(im))


This is my current code

https://gist.github.com/xellDart/e8760c6d7be1dd2a7090ea736dfc1d69

**Comment by sctrueew** [comment]
How to run the classification model in ONNX C++?

**Comment by HonestyBrave** [comment]
@xellDart hello, have you solve the problem? i also have the question
